05.wc/wc.rb

```diff
@@ -0,0 +1,78 @@
+# ! /usr/bin/env ruby
+# frozen_string_literal: true
+
+require 'optparse'
+
+SEPARATOR_SIZE = 1
+STDIN_FILENAME = '-'
+TOTAL = '合計'
+
+def main
+  options = parse_options(ARGV)
+  rows = []
+  ARGF.each(nil) do |input_text|
+    counts = {}
+
+    counts[:filename] = ARGF.filename
+    counts[:line] = input_text.lines.count if options[:l]
+    counts[:word] = input_text.split(/\s+/).size if options[:w]
+    counts[:byte] = input_text.size if options[:c]
+    rows << counts
+  end
+  rows << calculate_total_rows(rows) if rows.size > 1
+  max_widths = calculate_max_widths(rows)
+  show_rows(rows, max_widths)
+end
+
+def calculate_total_rows(rows)
+  total = rows.inject({}) do |result, count|
+    result.merge(count) do |key, current_val, adding_value|
+      current_val + adding_value unless key == :filename
+    end
+  end
+  total[:filename] = TOTAL
+  total
+end
+
+def show_rows(rows, max_column_widths)
+  rows.each do |row|
+    columns = {}
+    row.each do |name, value|
+      columns[name] = if name == :filename
+                        value unless value == STDIN_FILENAME
+                      else
+                        column_width = max_column_widths[name]
+                        value.to_s.rjust(column_width) + ' ' * SEPARATOR_SIZE
+                      end
+    end
+    puts columns.values_at(:line, :word, :byte, :filename).join
+  end
+end
+
+def calculate_max_widths(rows)
+  max_widths = {}
+  keys = rows.map(&:keys).flatten.uniq
+  keys.each do |key|
```

Comment:
> もっと驚きの少ない実装がないか以下の二通りの方法を考えましたが

ここはもっとシンプルに考えていいと思います。

```suggestion
  %i[line word byte].each do |key|
```

こんな感じで十分かなと思ったのですが、どうでしょうか？

あと、eachの代わりにto_hを使うとこのメソッドの実装をよりシンプルにできそうです
https://docs.ruby-lang.org/ja/latest/method/Array/i/to_h.html